pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -296,24 +329,56 @@ func (d *defaultDriver) MaybeEnableShardsAllocation(
 	return results
 }
 
-func (ctx *rollingUpgradeCtx) prepareClusterForNodeRestart(esClient esclient.Client, esState ESState) error {
+func (ctx *rollingUpgradeCtx) readyToDelete(pod corev1.Pod) (bool, error) {
+	if !supportsNodeShutdown(ctx.esClient.Version()) {
+		return true, nil // always OK to restart pre node shutdown support
+	}
+	if pod.Status.Phase != corev1.PodRunning {
+		// there is no point in trying to query the shutdown status of a Pod that is not running
+		return true, nil
+	}
+	response, err := ctx.nodeShutdown.ShutdownStatus(ctx.parentCtx, pod.Name)
+	if err != nil {
+		return false, err
+	}
+	return response.Status == esclient.ShutdownComplete, nil
+}
+
+func (ctx *rollingUpgradeCtx) requestNodeRestarts(podsToRestart []corev1.Pod) error {
+	var podNames []string //nolint:prealloc
+	for _, p := range podsToRestart {
+		if p.Status.Phase != corev1.PodRunning {
+			// There is no point in trying to shut down a Pod that is not running.
+			// Basing this off of the cached Kubernetes client's world view opens up a few edge
+			// cases where a Pod might in fact already be running but the client's cache is not yet
+			// up to date. But the trade-off made here i.e. accepting an ungraceful shutdown in these
+			// edge case vs. being able to automatically unblock configuration rollouts that are blocked
+			// due to misconfiguration, for example unfulfillable node selectors, seems worth it.
+			continue
+		}
+		podNames = append(podNames, p.Name)
+	}
+	// Note that ReconcileShutdowns would cancel ongoing shutdowns when called with no podNames
+	// this is however not the case in the rolling upgrade logic where we exit early if no pod needs to be rotated.
+	return ctx.nodeShutdown.ReconcileShutdowns(ctx.parentCtx, podNames)
+}
+
+func (ctx *rollingUpgradeCtx) prepareClusterForNodeRestart(podsToUpgrade []corev1.Pod) error {
+	if supportsNodeShutdown(ctx.esClient.Version()) {
+		return ctx.requestNodeRestarts(podsToUpgrade)
+	}
 	// Disable shard allocations to avoid shards moving around while the node is temporarily down
-	shardsAllocationEnabled, err := esState.ShardAllocationsEnabled()
+	shardsAllocationEnabled, err := ctx.esState.ShardAllocationsEnabled()
 	if err != nil {
 		return err
 	}
 	if shardsAllocationEnabled {
 		log.Info("Disabling shards allocation", "es_name", ctx.ES.Name, "namespace", ctx.ES.Namespace)
-		if err := esClient.DisableReplicaShardsAllocation(ctx.parentCtx); err != nil {
+		if err := ctx.esClient.DisableReplicaShardsAllocation(ctx.parentCtx); err != nil {
 			return err
 		}
 	}
 
 	// Request a flush to optimize indices recovery when the node restarts.
-	if err := doFlush(ctx.parentCtx, ctx.ES, esClient); err != nil {
-		return err
-	}
-
-	// TODO: halt ML jobs on that node
```

Comment:
was this something we still needed to do; is there a backlog story for implementation?

Reply:
This was an unaddressed issue that the node shutdown API takes care of for all versions of Elasticsearch that support the API. We have no plans of managing ML jobs (or ILM policies which is another unsolved thing) for older versions of Elasticsearch.